src/main/java/org/junit/runners/BlockJUnit4ClassRunner.java

```diff
@@ -380,29 +382,32 @@ protected Statement withAfters(FrameworkMethod method, Object target,
                 target);
     }
 
-    private Statement withRules(FrameworkMethod method, Object target,
-            Statement statement) {
-        List<TestRule> testRules = getTestRules(target);
-        Statement result = statement;
-        result = withMethodRules(method, testRules, target, result);
-        result = withTestRules(method, testRules, result);
-
-        return result;
-    }
-
-    private Statement withMethodRules(FrameworkMethod method, List<TestRule> testRules,
-            Object target, Statement result) {
-        Statement withMethodRules = result;
-        for (org.junit.rules.MethodRule each : getMethodRules(target)) {
+    private Statement withRules(FrameworkMethod method, Object target, Statement statement) {
+        RuleContainer ruleContainer = new RuleContainer() {
+            protected Integer getOrder(FrameworkMember member) {
+                Rule rule = member.getAnnotation(Rule.class);
+                return rule != null ? rule.order() : null;
+            }
+        };
+        List<TestRule> testRules;
+        List<MethodRule> methodRules;
+        AnnotatedValueListener.CURRENT.set(ruleContainer);
```

Comment:
We would probably want to do something similar in `ParentRunner.withClassRules()`